This patch fixes a freeze present on Haswell machines with APTIO IV and probably some others.
Backported from legacy EFI Shell.

diff --git a/ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c b/ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c
index 322d510e..2cf23012 100644
--- a/ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c
+++ b/ShellPkg/Library/UefiShellLevel2CommandsLib/Load.c
@@ -36,6 +36,7 @@ ConnectAllEfi (
   UINTN       HandleCount;
   EFI_HANDLE  *HandleBuffer;
   UINTN       Index;
+  VOID        *DriverBinding;
 
   Status = gBS->LocateHandleBuffer (
                   AllHandles,
@@ -49,6 +50,19 @@ ConnectAllEfi (
   }
 
   for (Index = 0; Index < HandleCount; Index++) {
+    Status = gBS->HandleProtocol (
+      HandleBuffer[Index],
+      &gEfiDevicePathProtocolGuid,
+      &DriverBinding
+      );
+
+    if (EFI_ERROR (Status)) {
+      //
+      // Calling ConnectController on non-driver results in freezes on APTIO IV.
+      //
+      continue;
+    }
+
     Status = gBS->ConnectController (HandleBuffer[Index], NULL, NULL, TRUE);
   }
 
